### ISSUE TITLE
Feature: Support MRSimulation format

### ISSUE THREAD

**Issue by crtag** [atomCAD/atomCAD#35]
[Molecular Renderer](https://github.com/philipturner/molecular-renderer) produces simulation results for replay in the following format: [Codec/Plain Text Format](https://github.com/philipturner/molecular-renderer/blob/main/Documentation/MRSimulation.md#plain-text-format)
atomCAD will benefit from the capability to replay or import simulations generated by other tools.

The initial [pull request ](https://github.com/atomCAD/atomCAD/pull/32) to implement codec contains a detailed outline of the recommended approach to tackle rendering challenges.

This requires a number of new features developed and respective skills, so will need a number of contributors

1. Support codec (cleanup code base, optimize for Rust best practice) https://github.com/atomCAD/atomCAD/pull/40
2. Enable atom movements (each "frame" defines x,y,z for each atom)
3. Implement a new rendering mechanism (pipeline to feed simulation frames into display frames)
4. New UI components (generic import, play, pause controls)

**Comment by philipturner** [comment]
One issue, detailed extensively in the pull request, is whether we can use Vsync handlers to measure frames in perfect integer quantities. Example [here](https://github.com/philipturner/molecular-renderer/blob/ec7bfbd9540b4830e03c5d2cbf223247bc39171a/Sources/MolecularRendererApp/Rendering/VsyncHandler.swift).

Another idea is interpolating frames measured as floating-point numbers, using the two nearest integer multiples of the framerate. This would be a fallback on WASM when native Vsync APIs are unavailable.

**Comment by maaku** [comment]
Since my review has been requested on the linked PR, I'm copying over some relevant text that I posted on the Discord:

I'm not quite sure how this project came about, but I have never had any intention of adding support for MR into atomCAD, nor do I think making a specialized viewer for MR simulations runs to be within scope. I've gone through the github issue and pull request, but if there's a description of why this is being worked on, I've missed it. It just seems to be assumed as necessary, whereas from my perspective it's out of the blue.

MM4 is insufficient for our needs. atomCAD's approach will be to use the most accurate and general-purpose molecular dynamics codes (GFN-FF in most cases, ab initio methods when needed) to validate the properties of individual parts and part interfaces, then use continuum / finite element methods, or linear analysis to validate designs or simulate larger assemblies.

I don't know what use cases there are for generating very long simulations and then playing them back inside the editor, other than creating promotional material. And that seems like something that would be better done in blender.

If there is value in supporting some sort of simulation playback mode, and I still need to be convinced of that, then I would highly suggest looking into the already existing field of mesh compression, which exploits both spatial and temporal locality to compress animation data consisting of moving vertices, not unlike our moving atom centers. There are various lossless schemes, and these should yield much better performance and compression ratios than the MRSimulation format which is essentially gzip'd text. So I don't see the point of standardizing on a format that is, at best, a temporary stopgap solution before something better is implemented.

**Comment by philipturner** [comment]
We have some very obvious ideological differences, but I would hope we can put this aside. I am trying to get my own product to work on other platforms, and your code base seems like something that fits. In its current state, it has a production-quality imposter method renderer, designed for molecular nanotechnology, that can handle >100M atoms. I don't see that capability in any other software than my own. As an alternative, I could port my renderer to Vulkan and Linux. That approach would be more time-consuming than working with you and atomCAD, negotiating despite our differences.

My intention is to create a full-stack CAD workflow, _potentially/unfortunately competing with yours_, including a hardware description language, high-efficiency molecular simulation, and scaling to over 1 million atoms, _on the timescale of months_. I have been carefully building all the necessary software components for ~1 year, long before I met you. The first critical part was OpenMM, where I fixed some internal algorithmic issues causing $O(n^2)$ scaling at several million atoms. Next was a ray tracer than can handle over 1 million atoms. Not 100 million like yours, but still more than other apps, which aren't designed for high-throughput data-intensive tasks. Most other software is written in Python. I have seem Tom, a highly experienced nanomechanical engineer, admit to struggling from broken software not designed for the next generation of technology. Just like how AI software is fundamentally broken, and needs a first-principles rethink like Modular.

Being in Rust, your code base has proven to deliver more than outstanding performance. I made some benchmarks of a data deserializer in Python, Swift and Rust. Python delivered latencies unacceptable for my workflow, but the Rust version implemented by @crtag has latencies similar to Swift. This proves there is potential in your renderer, which nothing else out there is designed for. Something cross-platform, carefully designed over several years. After experimenting with different game engines for rendering, you went with an imposter method written from scratch in WGSL. WGSL and WebGPU are modern frameworks for rendering cross-platform, unlike OpenGL or CPU-based renderers in PyMOL. MSEP seems to have taken the same approach, but unfortunately they're closed source right now. I don't want to wait until they decide to be open-source.

About standardization, I do think some major pillars of this will need to be redone from scratch. That happened with my simulator, the HDL, and the MRSimulation format. All three are on "version 2" nearing completion soon. Even though "version 1" was a viable workflow. While the current state may be temporary in some respects, the end goal is a workflow that can scale to 1 million atom productive nanosystems. We can make zero assumptions about what will/won't be viable in the end goal, about which macroscale design/simulation techniques can be successfully transplanted to that paradigm. This is a new frontier in software technology, that I am glad to begin exploring with communities like the one you have created.

Sincerely,
Philip Turner
